api/server/v1/error.go

```diff
@@ -17,98 +17,292 @@ package api
 import (
 	"encoding/json"
 	"fmt"
+	"time"
 
-	"github.com/golang/protobuf/ptypes/any"
+	"github.com/golang/protobuf/proto"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
 	spb "google.golang.org/genproto/googleapis/rpc/status"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
-	"google.golang.org/protobuf/proto"
-	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/durationpb"
 )
 
+// This file contains helpers for error handling
+//
+// 1. GRPC interface
+// We reuse GRPCs standard payloads to propagate extended error information:
+// https://cloud.google.com/apis/design/errors
+// Our extended error code is passed in ErrorInfo and automatically unmarshalled
+// on the client.
+// The flow:
+//		* Server uses `api.Errorf({tigris code}, ...)` to report a TigrisError
+//      * TigrisError implements `GRPCStatus()` interface, so GRPC code can construct a GRPC error out of it
+//      * Client code calls `FromStatusError` to reconstruct TigrisError from GRPC status and it's payloads
+//        (Extended code is taken from errdetails.ErrorInfo and retry delay from errdetails.RetryInfo)
+//
+// 2. HTTP interface
+// So as HTTP interface is intended to be inspected by users, we marshal HTTP errors
+// in a more human-readable form, less verbose then GRPC errors.
+//
+// Example HTTP error:
+// {
+//   "error": {
+//      "code": "ALREADY_EXISTS"
+//      "message": "database already exists"
+//      "retry_delay": "1s"
+//   }
+// }
+//
+// The flow:
+//		* Server uses `api.Errorf({tigris code}, ...)` to report a TigrisError
+//      * We provide TigrisError.As(*runtime.HTTPStatusError) to be able to override HTTP
+//        status code in GRPC gateway if need to
+//      * The error is converted to GRPC status by GRPCStatus() interface
+//      * We setup (see CustomMarshaller) and provide custom marshaller of
+//        the GRPC status (see MarshalStatus) in GRPC gateway
+//      * Client parses the error using UnmarshalStatus into TigrisError
+
 // TigrisError is our Tigris HTTP counterpart of grpc status. All the APIs should use this Error to return as a user facing
 // error. TigrisError will return grpcStatus to the muxer so that grpc client can see grpcStatus as the final output. For
 // HTTP clients see the **MarshalStatus** method where we are returning the response by not the grpc code as that is not
 // needed for HTTP clients.
 type TigrisError struct {
 	// The status code, which should be an enum value of [google.rpc.Code][google.rpc.Code]. We don't need to marshal
 	// this code for HTTP clients.
-	Code codes.Code `json:"code,omitempty"`
+	Code Code `json:"code,omitempty"`
 	// A developer-facing error message.
 	Message string `json:"message,omitempty"`
-	// A list of messages that carry the error details. This is mainly to send our internal codes and messages.
-	Details []*ErrorDetails `json:"details,omitempty"`
```

Comment:
any reason of changing it?

Reply:
Yes, now it contains standard GRPC error details payloads. errdetails.ErrorInfo, errdetails.RetryInfo.